<!-- This comment will put IE 6, 7 and 8 in quirks mode -->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>cgatools: Class List</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javaScript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body onload='searchBox.OnSelectItem(0);'>
<!-- Generated by Doxygen 1.6.2 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <img id="MSearchSelect" src="search/search.png"
             onmouseover="return searchBox.OnSearchSelectShow()"
             onmouseout="return searchBox.OnSearchSelectHide()"
             alt=""/>
        <input type="text" id="MSearchField" value="Search" accesskey="S"
             onfocus="searchBox.OnSearchFieldFocus(true)" 
             onblur="searchBox.OnSearchFieldFocus(false)" 
             onkeyup="searchBox.OnSearchFieldChange(event)"/>
        <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
        </div>
      </li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li class="current"><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>Class List</h1>Here are the classes, structs, unions and interfaces with brief descriptions:<table>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1Allele.html">cgatools::variants::Allele</a></td><td class="indexvalue">A class that corresponds to a single allele within a locus in a Complete Genomics variant file </td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1variants_1_1AlleleDiffSegment.html">cgatools::variants::AlleleDiffSegment</a></td><td class="indexvalue">Structure to keep track of a segment the reference, as well as calls aligned to that segment for two calldiff files </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1AlleleSequenceAligner.html">cgatools::mapping::AlleleSequenceAligner</a></td><td class="indexvalue">The class is used to iterate through the bases of a variation allele from an evidenceInterval file operator[] returns either a base from the allele or a reference base if the index points outside of the allele range </td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1variants_1_1Call_1_1AlternativeCall.html">cgatools::variants::Call::AlternativeCall</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1Call_1_1AlternativeCallsField.html">cgatools::variants::Call::AlternativeCallsField</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1reference_1_1AmbiguousRegion.html">cgatools::reference::AmbiguousRegion</a></td><td class="indexvalue">Struct to describe an ambiguous region of reference, within a chromosome </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1BaseLaneBatchStreams.html">cgatools::mapping::BaseLaneBatchStreams</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1BaseMap2SamConverter.html">cgatools::mapping::BaseMap2SamConverter</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1CacheOutStreams.html">cgatools::mapping::CacheOutStreams</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1calib_1_1CalibratedScorer.html">cgatools::variants::calib::CalibratedScorer</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1variants_1_1Call.html">cgatools::variants::Call</a></td><td class="indexvalue">A struct that corresponds to a single line of a Complete Genomics variant file </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1CallDiff.html">cgatools::command::CallDiff</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1CallDiffResult.html">cgatools::variants::CallDiffResult</a></td><td class="indexvalue">Class to hold the result of diffing two superloci </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1VariantFileIterator_1_1CallFilter.html">cgatools::variants::VariantFileIterator::CallFilter</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1ListVariants_1_1CallOrder.html">cgatools::command::ListVariants::CallOrder</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1util_1_1FileSourceDevice_1_1category.html">cgatools::util::FileSourceDevice::category</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1util_1_1FileDescriptorDevice_1_1category.html">cgatools::util::FileDescriptorDevice::category</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1util_1_1FileSinkDevice_1_1category.html">cgatools::util::FileSinkDevice::category</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1CharField.html">cgatools::util::CharField</a></td><td class="indexvalue">A <a class="el" href="classcgatools_1_1util_1_1DelimitedFieldParser.html" title="Class that parses one field of a delimited line.">DelimitedFieldParser</a> that records a character (not an unsigned integer) </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1reference_1_1ChromosomeIdField.html">cgatools::reference::ChromosomeIdField</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1Cigar.html">cgatools::mapping::Cigar</a></td><td class="indexvalue">Provides support for manipulations on SAM/BAM and CGI Evidence CIGAR </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1Cigar_1_1CigarElement.html">cgatools::mapping::Cigar::CigarElement</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1cgdata_1_1CnvDetailData.html">cgatools::cgdata::CnvDetailData</a></td><td class="indexvalue">Per-window CNV data </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1CnvDetailStore.html">cgatools::cgdata::CnvDetailStore</a></td><td class="indexvalue">Loads CNV window data from a file and caches it in memory </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1copynumber_1_1CnvFileVcfRecordSource.html">cgatools::copynumber::CnvFileVcfRecordSource</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1copynumber_1_1CnvFileVcfRecordWriter.html">cgatools::copynumber::CnvFileVcfRecordWriter</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1cgdata_1_1CnvSegmentData.html">cgatools::cgdata::CnvSegmentData</a></td><td class="indexvalue">Per-segment CNV data </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1CnvSegmentStore.html">cgatools::cgdata::CnvSegmentStore</a></td><td class="indexvalue">Loads CNV segment data from a file and caches it in memory </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Command.html">cgatools::command::Command</a></td><td class="indexvalue">The base class for writing new commands </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1reference_1_1CompactDnaSequence.html">cgatools::reference::CompactDnaSequence</a></td><td class="indexvalue">Class to describe the DNA sequence of a chromosome, in a compact manner </td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1util_1_1GenericHistogram_1_1CompareBuckets.html">cgatools::util::GenericHistogram&lt; Value, Index &gt;::CompareBuckets</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1CompareBySide.html">cgatools::junctions::CompareBySide</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1CompareJunctionsBySide.html">cgatools::junctions::CompareJunctionsBySide</a></td><td class="indexvalue">Compares two junctions by a given side Uses the other side and then junction id as secondary keys </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1CompressedInputStream.html">cgatools::util::CompressedInputStream</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1CompressedOutputStream.html">cgatools::util::CompressedOutputStream</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1CounterSet.html">cgatools::junctions::CounterSet&lt; Fields, Type &gt;</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1calib_1_1CoverageBinner.html">cgatools::variants::calib::CoverageBinner</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Crr2Fasta.html">cgatools::command::Crr2Fasta</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1reference_1_1CrrFile.html">cgatools::reference::CrrFile</a></td><td class="indexvalue">Compact Randomly-accessible Reference file </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1reference_1_1CrrFileWriter.html">cgatools::reference::CrrFileWriter</a></td><td class="indexvalue">Writes a <a class="el" href="classcgatools_1_1reference_1_1CrrFile.html" title="Compact Randomly-accessible Reference file.">CrrFile</a> to an output stream </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1DecodeCrr.html">cgatools::command::DecodeCrr</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1DelimitedField.html">cgatools::util::DelimitedField</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1DelimitedFieldParser.html">cgatools::util::DelimitedFieldParser</a></td><td class="indexvalue">Class that parses one field of a delimited line </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1DelimitedFile.html">cgatools::util::DelimitedFile</a></td><td class="indexvalue">Used to parse files delimited in the way found in a typical Complete Genomics data file </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1DelimitedFileMetadata.html">cgatools::util::DelimitedFileMetadata</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1DelimitedLineParser.html">cgatools::util::DelimitedLineParser</a></td><td class="indexvalue">A class that parses delimited lines </td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1copynumber_1_1diploidData.html">cgatools::copynumber::diploidData</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1cgdata_1_1EvidenceReader_1_1DnbRecord.html">cgatools::cgdata::EvidenceReader::DnbRecord</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1DnbSequenceIterator.html">cgatools::mapping::DnbSequenceIterator</a></td><td class="indexvalue">An iterator through an aligned read sequence The iterator is able to return bases in reverse order or/and complimented </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1DnbStructure.html">cgatools::cgdata::DnbStructure</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Evidence2Cache.html">cgatools::command::Evidence2Cache</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Evidence2CacheConfig.html">cgatools::command::Evidence2CacheConfig</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Evidence2Sam.html">cgatools::command::Evidence2Sam</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Evidence2SamConfig.html">cgatools::command::Evidence2SamConfig</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1EvidenceCacheBuilder.html">cgatools::mapping::EvidenceCacheBuilder</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1EvidenceCacheDnbRecord.html">cgatools::mapping::EvidenceCacheDnbRecord</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1EvidenceCacheReader.html">cgatools::mapping::EvidenceCacheReader</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1EvidenceDnbRecord.html">cgatools::mapping::EvidenceDnbRecord</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1EvidenceReader.html">cgatools::cgdata::EvidenceReader</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1EvidenceSamRecord.html">cgatools::mapping::EvidenceSamRecord</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1Exception.html">cgatools::util::Exception</a></td><td class="indexvalue">A basic exception type </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Fasta2Crr.html">cgatools::command::Fasta2Crr</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1FastRangeSet.html">cgatools::util::FastRangeSet</a></td><td class="indexvalue">Class is designed to support a relatively small set of ranges and answer the questions:</p>
<ul>
<li>does a given range intersect one of the ranges in the <a class="el" href="classcgatools_1_1util_1_1FastRangeSet.html" title="the class is designed to support a relatively small set of ranges and answer the...">FastRangeSet</a>?</li>
<li>is a given location contained by one of the ranges in the <a class="el" href="classcgatools_1_1util_1_1FastRangeSet.html" title="the class is designed to support a relatively small set of ranges and answer the...">FastRangeSet</a>? </li>
</ul>
</td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1FileDescriptorDevice.html">cgatools::util::FileDescriptorDevice</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1util_1_1FileSinkDevice.html">cgatools::util::FileSinkDevice</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1util_1_1FileSourceDevice.html">cgatools::util::FileSourceDevice</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1FullGenomeJunctionComparator.html">cgatools::junctions::FullGenomeJunctionComparator</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1HalfDnbStructure_1_1Gap.html">cgatools::cgdata::HalfDnbStructure::Gap</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1GapAndConcordanceExtractor.html">cgatools::mapping::GapAndConcordanceExtractor</a></td><td class="indexvalue">The class scans a given alignment for a given allele, provided by <a class="el" href="classcgatools_1_1mapping_1_1AlleleSequenceAligner.html" title="The class is used to iterate through the bases of a variation allele from an evidenceInterval...">AlleleSequenceAligner</a> in the constructor </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1GapProbabilityAndConcordanceExtractor.html">cgatools::mapping::GapProbabilityAndConcordanceExtractor</a></td><td class="indexvalue">The class extends <a class="el" href="classcgatools_1_1mapping_1_1GapAndConcordanceExtractor.html" title="The class scans a given alignment for a given allele, provided by AlleleSequenceAligner...">GapAndConcordanceExtractor</a> by computing the probablility of sequence-dependent gaps in addition to the metrics computed by the parent class </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1GapEst_1_1GapsEstimator.html">cgatools::mapping::GapEst::GapsEstimator</a></td><td class="indexvalue">Estimates wobble (small) gap probabilities, based on the sequence at a particular offset from the end of the clone </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1GapTable_1_1GapsRecord.html">cgatools::cgdata::GapTable&lt; T &gt;::GapsRecord</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1GapTable.html">cgatools::cgdata::GapTable&lt; T &gt;</a></td><td class="indexvalue">The class contains simple gap estimation table template </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1reference_1_1GeneDataStore.html">cgatools::reference::GeneDataStore</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1reference_1_1GeneDescriptionData.html">cgatools::reference::GeneDescriptionData</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1GenericHistogram.html">cgatools::util::GenericHistogram&lt; Value, Index &gt;</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1GenomeMetadata.html">cgatools::cgdata::GenomeMetadata</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1reference_1_1GetRangeBoundary.html">cgatools::reference::GetRangeBoundary</a></td><td class="indexvalue">Class that can be used as GetBoundary template parameter for an IntervalTree built on <a class="el" href="classcgatools_1_1reference_1_1Range.html" title="Class representing a range of sequence in a CrrFile.">reference::Range</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1HalfDnbStructure.html">cgatools::cgdata::HalfDnbStructure</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1IgnoreField.html">cgatools::util::IgnoreField</a></td><td class="indexvalue">A no-op <a class="el" href="classcgatools_1_1util_1_1DelimitedFieldParser.html" title="Class that parses one field of a delimited line.">DelimitedFieldParser</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1IndirectComparator.html">cgatools::util::IndirectComparator&lt; Container &gt;</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1InLaneBatchStreams.html">cgatools::mapping::InLaneBatchStreams</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1InputStream.html">cgatools::util::InputStream</a></td><td class="indexvalue">A std::istream with good error messages </td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1cgdata_1_1EvidenceReader_1_1IntervalRecord.html">cgatools::cgdata::EvidenceReader::IntervalRecord</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1IntervalTree.html">cgatools::util::IntervalTree&lt; TRange, TBoundary, TValue, Overlap, GetBoundary, BoundaryLess &gt;</a></td><td class="indexvalue">Class that maintains a mapping from key to value, where the key is a type with range semantics, for the purpose of finding the intersection of a range to a set of ranges </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Join.html">cgatools::command::Join</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1JuctionStatisticsCollector.html">cgatools::junctions::JuctionStatisticsCollector</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1Junction.html">cgatools::junctions::Junction</a></td><td class="indexvalue">Main class describing a single <a class="el" href="classcgatools_1_1junctions_1_1Junction.html" title="Main class describing a single Junction.">Junction</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Junction2Vcf.html">cgatools::command::Junction2Vcf</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1JunctionCmp.html">cgatools::command::JunctionCmp</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1JunctionCmpConfig.html">cgatools::command::JunctionCmpConfig</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1JunctionCounterFields.html">cgatools::junctions::JunctionCounterFields</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1JunctionDiff.html">cgatools::command::JunctionDiff</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1JunctionDiffConfig.html">cgatools::command::JunctionDiffConfig</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1JunctionFile.html">cgatools::junctions::JunctionFile</a></td><td class="indexvalue">Class describing a <a class="el" href="classcgatools_1_1junctions_1_1Junction.html" title="Main class describing a single Junction.">Junction</a> file </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1JunctionRef.html">cgatools::junctions::JunctionRef</a></td><td class="indexvalue">A utility class used as a lightweight reference to a junction record and might contain a list of references to associated junctions </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1JunctionRefs.html">cgatools::junctions::JunctionRefs</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1junctions_1_1JunctionRefSide.html">cgatools::junctions::JunctionRefSide</a></td><td class="indexvalue">One junction side, sortable in the reference order </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1JunctionSideSection.html">cgatools::junctions::JunctionSideSection</a></td><td class="indexvalue">Class describing the left or right side of a junction </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1JunctionType.html">cgatools::command::JunctionType</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1JunctionVcfRecordSource.html">cgatools::junctions::JunctionVcfRecordSource</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1JunctionVcfRecordWriter.html">cgatools::junctions::JunctionVcfRecordWriter</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1junctions_1_1JunctionVcfWriter.html">cgatools::junctions::JunctionVcfWriter</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1LibraryData.html">cgatools::mapping::LibraryData</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1LibraryMetadata.html">cgatools::cgdata::LibraryMetadata</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1LibraryMetadataContainer.html">cgatools::mapping::LibraryMetadataContainer</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1LibRecord.html">cgatools::cgdata::LibRecord</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1ListCrr.html">cgatools::command::ListCrr</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1ListVariants.html">cgatools::command::ListVariants</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1reference_1_1Location.html">cgatools::reference::Location</a></td><td class="indexvalue">Class representing a location in a <a class="el" href="classcgatools_1_1reference_1_1CrrFile.html" title="Compact Randomly-accessible Reference file.">CrrFile</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1Locus.html">cgatools::variants::Locus</a></td><td class="indexvalue">A class that corresponds to a single locus in a Complete Genomics variant file </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1LocusAnnotations.html">cgatools::variants::LocusAnnotations</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Map2Sam.html">cgatools::command::Map2Sam</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1Map2SamConfig.html">cgatools::mapping::Map2SamConfig</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1Map2SamConverter.html">cgatools::mapping::Map2SamConverter</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1Md5Context.html">cgatools::util::Md5Context</a></td><td class="indexvalue">A class to compute the MD5 hash of a sequence of bytes </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1Md5Digest.html">cgatools::util::Md5Digest</a></td><td class="indexvalue">The MD5 result type </td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1mobileelement_1_1meiData.html">cgatools::mobileelement::meiData</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mobileelement_1_1MeiFileVcfRecordSource.html">cgatools::mobileelement::MeiFileVcfRecordSource</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mobileelement_1_1MeiFileVcfRecordWriter.html">cgatools::mobileelement::MeiFileVcfRecordWriter</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1MergedMap2Sam.html">cgatools::command::MergedMap2Sam</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1MergedMap2SamConfig.html">cgatools::mapping::MergedMap2SamConfig</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1MergedMap2SamConverter.html">cgatools::mapping::MergedMap2SamConverter</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1reference_1_1RangeAnnotationStore_1_1MidpointField.html">cgatools::reference::RangeAnnotationStore&lt; Derived, TValue &gt;::MidpointField&lt; Value &gt;</a></td><td class="indexvalue">Special delimited field parser to put same column ("position") in two places </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1MkVcf.html">cgatools::command::MkVcf</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1GapEst_1_1NoCallSequenceRetriever.html">cgatools::mapping::GapEst::NoCallSequenceRetriever</a></td><td class="indexvalue">Retrieves all no-calls </td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1copynumber_1_1nondiploidData.html">cgatools::copynumber::nondiploidData</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1OutLaneBatchStreams.html">cgatools::mapping::OutLaneBatchStreams</a></td><td class="indexvalue">Creates a family of lane/batch files and generates standard metadata header </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamRecordGenerator_1_1OutputFileDescriptor.html">cgatools::mapping::SamRecordGenerator::OutputFileDescriptor</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1OutputStream.html">cgatools::util::OutputStream</a></td><td class="indexvalue">Std::ostream with good error messages, and which does fsync on close on UNIX </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1PhasedAllele.html">cgatools::variants::PhasedAllele</a></td><td class="indexvalue">Class to hold allele-specific information for a <a class="el" href="classcgatools_1_1variants_1_1PhasedHypothesis.html" title="Class to hold a single possible phasing of loci in a Superlocus.">PhasedHypothesis</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1PhasedHypothesis.html">cgatools::variants::PhasedHypothesis</a></td><td class="indexvalue">Class to hold a single possible phasing of loci in a <a class="el" href="classcgatools_1_1variants_1_1Superlocus.html" title="The Superlocus class.">Superlocus</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1reference_1_1Range.html">cgatools::reference::Range</a></td><td class="indexvalue">Class representing a range of sequence in a <a class="el" href="classcgatools_1_1reference_1_1CrrFile.html" title="Compact Randomly-accessible Reference file.">CrrFile</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1reference_1_1RangeAnnotationStore.html">cgatools::reference::RangeAnnotationStore&lt; Derived, TValue &gt;</a></td><td class="indexvalue">Front end to IntervalTree based on reference ranges that allows to load annotations from a file </td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1reference_1_1RangeOverlap.html">cgatools::reference::RangeOverlap</a></td><td class="indexvalue">Class that can be used as Overlap template parameter for an IntervalTree built on <a class="el" href="classcgatools_1_1reference_1_1Range.html" title="Class representing a range of sequence in a CrrFile.">reference::Range</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1FastRangeSet_1_1RangeSet.html">cgatools::util::FastRangeSet::RangeSet</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamFileHeader_1_1Record.html">cgatools::mapping::SamFileHeader::Record</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1cgdata_1_1ReferenceSupportReader.html">cgatools::cgdata::ReferenceSupportReader</a></td><td class="indexvalue">Sequential reader for CGI reference support files provided as a part of the genome data package </td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1reference_1_1RepeatMaskerAnnotation.html">cgatools::reference::RepeatMaskerAnnotation</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1reference_1_1RepeatMaskerStore.html">cgatools::reference::RepeatMaskerStore</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Sam2Reads.html">cgatools::command::Sam2Reads</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1Sam2ReadsConfig.html">cgatools::command::Sam2ReadsConfig</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamFileHeader.html">cgatools::mapping::SamFileHeader</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamFileHeaderBlock.html">cgatools::mapping::SamFileHeaderBlock</a></td><td class="indexvalue">Used to manage sections of the SAM file header use "operator &lt;&lt;" to serialize the content into a SAM file </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamFileParser.html">cgatools::mapping::SamFileParser</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamFileRecord.html">cgatools::mapping::SamFileRecord</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamGeneratorConfig.html">cgatools::mapping::SamGeneratorConfig</a></td><td class="indexvalue">Sam Generator configuration parameters </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamHeaderTag.html">cgatools::mapping::SamHeaderTag</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamRecord.html">cgatools::mapping::SamRecord</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamRecordGenerator.html">cgatools::mapping::SamRecordGenerator</a></td><td class="indexvalue">Generates a SAM file record based on a given mapping record The final logic defining what goes into the output file is defined here </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamSequenceSplitter.html">cgatools::mapping::SamSequenceSplitter</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamTag.html">cgatools::mapping::SamTag</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamTagField.html">cgatools::mapping::SamTagField&lt; SamTag &gt;</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1SamTags.html">cgatools::mapping::SamTags&lt; SamTag &gt;</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1GapEst_1_1SequenceRetriever.html">cgatools::mapping::GapEst::SequenceRetriever</a></td><td class="indexvalue">Retrieves sequence at a given offset from the end of the clone </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1SideField.html">cgatools::util::SideField</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1SimpleHistogram.html">cgatools::util::SimpleHistogram</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1SnpDiff.html">cgatools::command::SnpDiff</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1copynumber_1_1somaticData.html">cgatools::copynumber::somaticData</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1StrandField.html">cgatools::util::StrandField</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1StringField.html">cgatools::util::StringField</a></td><td class="indexvalue">A <a class="el" href="classcgatools_1_1util_1_1DelimitedFieldParser.html" title="Class that parses one field of a delimited line.">DelimitedFieldParser</a> that records a copy of the field </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1StringSet.html">cgatools::util::StringSet</a></td><td class="indexvalue"><a class="el" href="classcgatools_1_1util_1_1StringSet.html" title="StringSet convenience class.">StringSet</a> convenience class </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1Superlocus.html">cgatools::variants::Superlocus</a></td><td class="indexvalue">The <a class="el" href="classcgatools_1_1variants_1_1Superlocus.html" title="The Superlocus class.">Superlocus</a> class </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1SuperlocusIterator.html">cgatools::variants::SuperlocusIterator</a></td><td class="indexvalue">A class to iterate through one or more variant files at a time, enabling access to a stream of <a class="el" href="classcgatools_1_1variants_1_1Superlocus.html" title="The Superlocus class.">Superlocus</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1mapping_1_1GapEst_1_1TemplateSequenceRetriever.html">cgatools::mapping::GapEst::TemplateSequenceRetriever&lt; TSequence &gt;</a></td><td class="indexvalue">Retrieves sequence at a given offset from the end of the clone </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1TestVariants.html">cgatools::command::TestVariants</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1Call_1_1TotalScoreField.html">cgatools::variants::Call::TotalScoreField</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1ValueField.html">cgatools::util::ValueField&lt; Value &gt;</a></td><td class="indexvalue">A <a class="el" href="classcgatools_1_1util_1_1DelimitedFieldParser.html" title="Class that parses one field of a delimited line.">DelimitedFieldParser</a> that uses <a class="el" href="namespacecgatools_1_1util.html#a91bd51f7a2cbe736cf441d6abc10f824" title="Parse the string to the given Value type.">cgatools::util::parseValue</a> to parse its value </td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1util_1_1ValueParser.html">cgatools::util::ValueParser&lt; Value, IsSigned &gt;</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1util_1_1ValueParser_3_01double_00_01true_01_4.html">cgatools::util::ValueParser&lt; double, true &gt;</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1util_1_1ValueParser_3_01float_00_01true_01_4.html">cgatools::util::ValueParser&lt; float, true &gt;</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1util_1_1ValueParser_3_01Value_00_01false_01_4.html">cgatools::util::ValueParser&lt; Value, false &gt;</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1util_1_1ValueParser_3_01Value_00_01true_01_4.html">cgatools::util::ValueParser&lt; Value, true &gt;</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1util_1_1ValueVectorField.html">cgatools::util::ValueVectorField&lt; Value &gt;</a></td><td class="indexvalue"><a class="el" href="classcgatools_1_1util_1_1DelimitedFieldParser.html" title="Class that parses one field of a delimited line.">DelimitedFieldParser</a> that parses a variable length list of values separated by its own delimiter (distinct from the field delimiter) </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1VarFileCombine.html">cgatools::command::VarFileCombine</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1command_1_1VarFilter.html">cgatools::command::VarFilter</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1Call_1_1VarFilterField.html">cgatools::variants::Call::VarFilterField</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1VariantFileIterator.html">cgatools::variants::VariantFileIterator</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1VariantFileVcfRecordSource.html">cgatools::variants::VariantFileVcfRecordSource</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1VariantFileVcfRecordWriter.html">cgatools::variants::VariantFileVcfRecordWriter</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1variants_1_1Call_1_1VarQualityField.html">cgatools::variants::Call::VarQualityField</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1variants_1_1VariantFileVcfRecordWriter_1_1VcfData.html">cgatools::variants::VariantFileVcfRecordWriter::VcfData</a></td><td class="indexvalue"></td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1conv_1_1VcfKvHeaderRecord.html">cgatools::conv::VcfKvHeaderRecord</a></td><td class="indexvalue">Struct to represent metadata records like: #source_KEY=VALUE </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1conv_1_1VcfRecordSource.html">cgatools::conv::VcfRecordSource</a></td><td class="indexvalue">Abstract base class that provides a stream of <a class="el" href="classcgatools_1_1conv_1_1VcfRecordWriter.html" title="Abstract base class that writes a VCF record.">VcfRecordWriter</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classcgatools_1_1conv_1_1VcfRecordWriter.html">cgatools::conv::VcfRecordWriter</a></td><td class="indexvalue">Abstract base class that writes a VCF record </td></tr>
  <tr><td class="indexkey"><a class="el" href="structcgatools_1_1conv_1_1VcfSubFieldHeaderRecord.html">cgatools::conv::VcfSubFieldHeaderRecord</a></td><td class="indexvalue"></td></tr>
</table>
</div>
<!--- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&nbsp;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&nbsp;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&nbsp;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&nbsp;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&nbsp;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&nbsp;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&nbsp;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&nbsp;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&nbsp;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&nbsp;</span>Defines</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<hr size="1"/><address style="text-align: right;"><small>Generated by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.6.2 </small></address>
</body>
</html>
